---
title: "d0-iPSC-analysis"
output: html_document
editor_options: 
  chunk_output_type: console
---

# Initial analysis of d0 iPSC large screen

```{r}
library(hdf5r)
library(sceptre)
library(Seurat)
library(ggplot2)
```


```{r}
metadata_folder = "./metadata/"
pairs_folder = "./results/2025-09-02/"
```


```{r}
# Helper to read a 10x h5 that may have multiple modalities and/or nested genome lists
read_10x_rna_grna <- function(h5, sample_name) {
  x <- Seurat::Read10X_h5(h5, use.names = TRUE)
  # If x is nested by genome (list of lists), flatten one level
  if (is.list(x) && !inherits(x, "dgCMatrix") && all(vapply(x, is.list, logical(1)))) {
    x <- unlist(x, recursive = FALSE)
  }
  if (inherits(x, "dgCMatrix")) {
    # single-modality file
    obj <- Seurat::CreateSeuratObject(counts = x, project = sample_name)
    return(obj)
  }
  # Identify modality names case-insensitively
  nms <- names(x)
  pick_name <- function(cands) {
    hit <- nms[match(tolower(cands), tolower(nms))]
    hit[!is.na(hit)][1]
  }
  rna_name  <- pick_name(c("Gene Expression","gene expression","GeneExpression","RNA"))
  grna_name <- pick_name(c("CRISPR Guide Capture","CRISPR Guides","CRISPR","Guide Capture"))

  if (is.na(rna_name)) stop("Could not find an RNA modality in: ", paste(nms, collapse=", "))

  obj <- Seurat::CreateSeuratObject(counts = x[[rna_name]], project = sample_name)
  if (!is.na(grna_name)) {
    obj[["gRNA"]] <- Seurat::CreateAssayObject(counts = x[[grna_name]])
  } else {
    message("No CRISPR/gRNA modality found in: ", h5)
  }
  obj
}
```

## Load the guides annotations
```{r}
guides_metadata = read.csv(paste0(metadata_folder,"All_guides_with_addnl_controls_altTSS.csv"))
guides_metadata
```

## Load the genes annotations
```{r}
genes_metadata = read.csv(paste0(metadata_folder,"D0-D2_screen_final_311_gene-primers-ordered.xlsx - Inner-311.csv"))
# Strip the Sequence of 	GTCTCGTGGGCTCGGAGATGTGTATAAGAGACA
genes_metadata$Inner = gsub("GTCTCGTGGGCTCGGAGATGTGTATAAGAGACA", "", genes_metadata$Sequence)

genes_annotations = read.csv(paste0(metadata_folder,"Final_gene_list_to_design.csv"))

# Add the gene type to the genes metadata dataframe
genes_metadata = merge(genes_metadata, genes_annotations[,c("gene","type")], by.x="Name", by.y="gene", all.x=TRUE)

# Clean up
rm(genes_annotations)

genes_metadata
```

## Read the count matrices in Seurat
```{r}
#h5_files_all <- list.files(path = ".", pattern = "*.h5", full.names = TRUE, recursive = TRUE)
h5_files_all <- list.files(path = "./data", pattern = "*.h5", full.names = TRUE, recursive = TRUE)
h5_files_all
sample_to_analyze = "100-D0-r2-A"
h5_files <- h5_files_all[grepl(sample_to_analyze, h5_files_all, fixed = TRUE)]
h5_files
#h5_sample_names <- gsub("(.*)_sample_filtered_feature_bc_matrix.h5", "\\1", h5_files)
h5_sample_names <- sub(".*/(.*)_sample_filtered_feature_bc_matrix\\.h5$", "\\1", h5_files)

h5_sample_names

seurat_objects_list <- mapply(function(data, sample_name) {
  read_10x_rna_grna(data, sample_name)
}, data = h5_files, sample_name = h5_sample_names, SIMPLIFY = FALSE)

names(seurat_objects_list) <- h5_sample_names

d0_data <- merge(x = seurat_objects_list[[1]], 
                 y = seurat_objects_list[2:length(seurat_objects_list)],
                 merge.data = TRUE,
                 add.cell.ids = h5_sample_names)
Assays(d0_data)
```

## QC the RNAseq
```{r}
# Checking gene presence
present <- intersect(genes_metadata$Name, rownames(d0_data))
# Get the list of genes for which there is no expression
missing <- setdiff(unique(genes_metadata$Name), present)
if (length(missing) > 0){
  warning("Missing values detected for " , length(missing), " genes\n")
  warning("Missing genes: ", paste(missing, collapse = ", "), "\n")
}else{
  message("All genes present in the dataset\n")
}

# Find gene type using the genes metadata
genes_type <- setNames(genes_metadata$type, genes_metadata$Name)
table(genes_type[missing])
```

```{r}
# Compute percentage of Mitochondrial genes
d0_data[["percent.mt"]] <- PercentageFeatureSet(d0_data, pattern = "^MT-", assay = "RNA")

#Remove the last character from orig.ident
d0_data$orig.dose.chip <- gsub(".$", "", d0_data$orig.ident)
table(d0_data$orig.dose.chip)

# Compute percent of UMIs in target genes
layers <- grep("^counts", Layers(d0_data[["RNA"]]), value = TRUE)
target_genes <- genes_metadata$Name

pct_list <- lapply(layers, function(ly) {
  M <- GetAssayData(d0_data, assay = "RNA", layer = ly)
  present <- intersect(target_genes, rownames(M))
  tot <- Matrix::colSums(M)
  tgt <- if (length(present)) Matrix::colSums(M[present, , drop = FALSE]) else rep(0, length(tot))
  setNames(ifelse(tot > 0, 100 * tgt / tot, 0), colnames(M))
})

# combine and align to all cells in object
pct_vec <- do.call(c, pct_list)
pct_vec <- pct_vec[colnames(d0_data)]

d0_data$percent_target_umis <- pct_vec

d0_data_filtered = subset(d0_data, subset = percent_target_umis >= 60)
```

Use ENSEMBLE ID as rownames

```{r}
x_names <- Read10X_h5(h5_files[1], use.names = TRUE)   # rownames = gene symbols
x_ids   <- Read10X_h5(h5_files[1], use.names = FALSE)  # rownames = Ensembl IDs

# Handle single-modality vs multi-modality (list)
pick_mod <- function(x, mod = c("Gene Expression","RNA")) {
  if (inherits(x, "dgCMatrix")) return(x)
  nms <- names(x); i <- match(tolower(mod), tolower(nms)); i <- i[!is.na(i)][1]
  if (is.na(i)) stop("Could not find RNA modality. Found: ", paste(nms, collapse=", "))
  x[[i]]
}

rna_names <- pick_mod(x_names)
rna_ids   <- pick_mod(x_ids)

# Build a mapping data.frame
gene_map <- data.frame(
  symbol  = rownames(rna_names),
  ensembl = rownames(rna_ids),
  stringsAsFactors = FALSE
)

rownames(d0_data_filtered)
d0_data_filtered[["RNA"]]@meta.data$ensembl = gene_map$ensembl[match(rownames(d0_data_filtered), gene_map$symbol)]
d0_data_filtered[["RNA"]]@meta.data$symbol = rownames(d0_data_filtered)
rownames(d0_data_filtered) = d0_data_filtered[["RNA"]]@meta.data$ensembl
```


### Sceptre analysis

```{r}
# Prepare the data for sceptre
#grna_target_df = read.table("metadata/gRNA-target.tsv", header = TRUE, sep = "\t")
grna_target_df = read.table(paste0(pairs_folder,"gRNA-target.tsv"), header = TRUE, sep = "\t")
head(grna_target_df)
colnames(grna_target_df) = c("grna_id", "grna_target")
grna_target_df$grna_id =  gsub("_", "-", grna_target_df$grna_id)
discovery_pairs = read.table(paste0(pairs_folder,"discovery_pairs_noposctrl.tsv"), header = TRUE, sep = "\t")
head(discovery_pairs)
positive_controls = read.table(paste0(pairs_folder,"positive_pairs.tsv"), header = FALSE, sep = "\t")
colnames(positive_controls) = c("grna_target", "response_id")
head(positive_controls)

d0_rna_assay = JoinLayers(d0_data_filtered[["RNA"]], new.layer.name = "counts", layers = "counts")
rna_counts <- GetAssayData(d0_rna_assay, assay = "RNA", layer = "counts")
grna_counts <- GetAssayData(d0_data_filtered, assay = "gRNA", layer = "counts")
grna_counts <- grna_counts[grna_target_df$grna_id, , drop = FALSE]
sceptre_object = sceptre::import_data(response_matrix = rna_counts[gene_map$ensembl[match(present, gene_map$symbol)], , drop = FALSE],
                                      grna_matrix = grna_counts,
                                      grna_target_data_frame = grna_target_df,
                                      moi = "high"
                                      )
sceptre_object
```

Quick clean-up

```{r}
rm(seurat_objects_list)
rm(d0_rna_assay)
rm(temp)
rm(d0_data)
rm(rna_counts)
rm(grna_counts)
rm(x_ids)
rm(x_names)
rm(rna_ids)
rm(rna_names)
gc()
```

QC and assign guides

```{r}
discovery_pairs = discovery_pairs[discovery_pairs$response_id %in% gene_map$ensembl[match(present, gene_map$symbol)], ]
positive_controls = positive_controls[positive_controls$response_id %in% gene_map$ensembl[match(present, gene_map$symbol)], ]

sceptre_object <- set_analysis_parameters(
  sceptre_object = sceptre_object,
  discovery_pairs = discovery_pairs,
  positive_control_pairs = positive_controls,
  side = "both"
)
plot_grna_count_distributions(sceptre_object)
sceptre_object <- assign_grnas(sceptre_object = sceptre_object,
                               method = "thresholding",
                               threshold = 10,
                               parallel = TRUE,
                               n_processors = 2)
sceptre_object <- run_qc(sceptre_object, p_mito_threshold = 0.2)
print(sceptre_object)
plot(sceptre_object)
plot_assign_grnas(sceptre_object)

```

Calibration check

```{r}
sceptre_object <- run_calibration_check(sceptre_object,
                                        parallel = TRUE,
                                        n_processors = 2,
                                        print_progress = TRUE,
                                        )
print(sceptre_object) # output suppressed for brevity
plot(sceptre_object)
```

Power check

```{r}
sceptre_object <- run_power_check(sceptre_object,
                                  parallel = TRUE,
                                  print_progress = TRUE,
                                  n_processors = 2)
print(sceptre_object) # output suppressed for brevity
plot(sceptre_object)

write.table(sceptre_object@power_result,
            file = paste0(sample_to_analyze, "_power_analysis.tsv"),
            sep = "\t",
            row.names = FALSE, 
            col.names = TRUE,
            quote = FALSE
            )

```


## Results by 10X chip and dose

```{r}
# Run sceptre for each chip and dose combination
dose.chip.samples = unique(d0_data_filtered$orig.dose.chip)

# Save all the objects in a list
sceptre_objects_list = list()
for (sample in dose.chip.samples[1]){
  message("Processing sample: ", sample)
  temp = subset(d0_data_filtered, subset = orig.dose.chip == sample)
  d0_rna_assay = JoinLayers(temp[["RNA"]], new.layer.name = "counts")
  rna_counts <- GetAssayData(d0_rna_assay, assay = "RNA", layer = "counts")
  grna_counts <- GetAssayData(temp, assay = "gRNA", layer = "counts")
  grna_counts <- grna_counts[grna_target_df$grna_id, , drop = FALSE]
  # Update rownames with ENSEMBL IDs
  
  sceptre_object = sceptre::import_data(response_matrix = rna_counts[gene_map$ensembl[match(present, gene_map$symbol)], , drop = FALSE],
                                        grna_matrix = grna_counts,
                                        grna_target_data_frame = grna_target_df,
                                        moi = "high"
                                        )
  # Keep in the discovery and positive pairs only the genes present in the matrix
  discovery_pairs = discovery_pairs[discovery_pairs$response_id %in% gene_map$ensembl[match(present, gene_map$symbol)], ]
  positive_controls = positive_controls[positive_controls$V2 %in% gene_map$ensembl[match(present, gene_map$symbol)], ]
  sceptre_object <- set_analysis_parameters(
    sceptre_object = sceptre_object,
    discovery_pairs = discovery_pairs,
    positive_control_pairs = positive_controls,
    side = "both"
  )
  sceptre_object <- assign_grnas(sceptre_object = sceptre_object, parallel = TRUE, n_processors = 4)
  sceptre_objects_list[[sample]] = sceptre_object
}

```


