---
title: "d0-iPSC-analysis"
output: html_document
editor_options: 
  chunk_output_type: console
---

# Initial analysis of d0 iPSC large screen

```{r}
library(hdf5r)
library(sceptre)
library(Seurat)
library(ggplot2)
```


```{r}
metadata_folder = "./metadata/"
pairs_folder = "./metadata/"
```


```{r}
# Helper to read a 10x h5 that may have multiple modalities and/or nested genome lists
read_10x_rna_grna <- function(h5, sample_name) {
  x <- Seurat::Read10X_h5(h5, use.names = TRUE)
  # If x is nested by genome (list of lists), flatten one level
  if (is.list(x) && !inherits(x, "dgCMatrix") && all(vapply(x, is.list, logical(1)))) {
    x <- unlist(x, recursive = FALSE)
  }
  if (inherits(x, "dgCMatrix")) {
    # single-modality file
    obj <- Seurat::CreateSeuratObject(counts = x, project = sample_name)
    return(obj)
  }
  # Identify modality names case-insensitively
  nms <- names(x)
  pick_name <- function(cands) {
    hit <- nms[match(tolower(cands), tolower(nms))]
    hit[!is.na(hit)][1]
  }
  rna_name  <- pick_name(c("Gene Expression","gene expression","GeneExpression","RNA"))
  grna_name <- pick_name(c("CRISPR Guide Capture","CRISPR Guides","CRISPR","Guide Capture"))

  if (is.na(rna_name)) stop("Could not find an RNA modality in: ", paste(nms, collapse=", "))

  obj <- Seurat::CreateSeuratObject(counts = x[[rna_name]], project = sample_name)
  if (!is.na(grna_name)) {
    obj[["gRNA"]] <- Seurat::CreateAssayObject(counts = x[[grna_name]])
  } else {
    message("No CRISPR/gRNA modality found in: ", h5)
  }
  obj
}
```

## Load the guides annotations
```{r}
guides_metadata = read.csv(paste0(metadata_folder,"All_guides_with_addnl_controls_altTSS.csv"))
guides_metadata
```

## Load the genes annotations
```{r}
genes_metadata = read.csv(paste0(metadata_folder,"D0-D2_screen_final_311_gene-primers-ordered.xlsx - Inner-311.csv"))
# Strip the Sequence of 	GTCTCGTGGGCTCGGAGATGTGTATAAGAGACA
genes_metadata$Inner = gsub("GTCTCGTGGGCTCGGAGATGTGTATAAGAGACA", "", genes_metadata$Sequence)

genes_annotations = read.csv(paste0(metadata_folder,"Final_gene_list_to_design.csv"))

# Add the gene type to the genes metadata dataframe
genes_metadata = merge(genes_metadata, genes_annotations[,c("gene","type")], by.x="Name", by.y="gene", all.x=TRUE)

# Clean up
rm(genes_annotations)

genes_metadata
```

## Read the count matrices in Seurat
```{r}
h5_files <- list.files(path = ".", pattern = "*.h5", full.names = TRUE, recursive = TRUE)
#h5_files_all <- list.files(path = "./data", pattern = "*.h5", full.names = TRUE, recursive = TRUE)
h5_files
#h5_files <- h5_files_all[grepl("100-D0-r2-A", h5_files_all, fixed = TRUE)]

h5_sample_names <- gsub("(.*)_sample_filtered_feature_bc_matrix.h5", "\\1", h5_files)
#h5_sample_names <- sub(".*/(.*)_sample_filtered_feature_bc_matrix\\.h5$", "\\1", h5_files)

h5_sample_names

seurat_objects_list <- mapply(function(data, sample_name) {
  read_10x_rna_grna(data, sample_name)
}, data = h5_files, sample_name = h5_sample_names, SIMPLIFY = FALSE)

names(seurat_objects_list) <- h5_sample_names

d0_data <- merge(x = seurat_objects_list[[1]], 
                 y = seurat_objects_list[2:length(seurat_objects_list)],
                 merge.data = TRUE,
                 add.cell.ids = h5_sample_names)
Assays(d0_data)
```

## QC the RNAseq
```{r}
plot(table(d0_data$orig.ident))

counts <- rev(table(d0_data$orig.ident))

op <- par(mar = c(4, 15, 2, 1))  # more room on the left for long labels
xmax <- max(counts)

bp <- barplot(
  counts,
  horiz = TRUE,
  las = 1,              
  xlab = "Number of barcode",
  cex.names = 1.5,
  xlim = c(0, xmax * 1.12)  
)
text(x = counts, y = bp, labels = counts, pos = 4, cex = 1.5, xpd = NA, offset = 0.5)

par(op)

# Checking gene presence
present <- intersect(genes_metadata$Name, rownames(d0_data))
# Get the list of genes for which there is no expression
missing <- setdiff(unique(genes_metadata$Name), present)
if (length(missing) > 0){
  warning("Missing values detected for " , length(missing), " genes\n")
  warning("Missing genes: ", paste(missing, collapse = ", "), "\n")
}else{
  message("All genes present in the dataset\n")
}

# Find gene type using the genes metadata
genes_type <- setNames(genes_metadata$type, genes_metadata$Name)
table(genes_type[missing])
```

```{r}
# Compute percentage of Mitochondrial genes
d0_data[["percent.mt"]] <- PercentageFeatureSet(d0_data, pattern = "^MT-", assay = "RNA")

#Remove the last character from orig.ident
d0_data$orig.dose.chip <- gsub(".$", "", d0_data$orig.ident)
table(d0_data$orig.dose.chip)

VlnPlot(d0_data, features = c("percent.mt"), ncol = 1, assay = "RNA")
VlnPlot(d0_data, features = c("percent.mt"), ncol = 1, assay = "RNA", group.by = "orig.dose.chip")

hist(d0_data$nCount_RNA,
     breaks = 30,
     main = "Number of UMIs per barcodes",
     sub = paste0("Total barcodes: ", ncol(d0_data)),
     xlab = "UMIs",
     ylab = "Number of barcodes",
     cex.main = 3,
     cex.lab = 1.5
     )
abline(v = median(d0_data$nCount_RNA), col = "red", lwd = 2)
# Label the median on the plot
text(median(d0_data$nCount_RNA) + 10000, 60000, labels = paste0("Median = ", round(median(d0_data$nCount_RNA), 2)), col = "red", cex = 1.5)

# Compute percent of UMIs in target genes
layers <- grep("^counts", Layers(d0_data[["RNA"]]), value = TRUE)
target_genes <- genes_metadata$Name

pct_list <- lapply(layers, function(ly) {
  M <- GetAssayData(d0_data, assay = "RNA", layer = ly)
  present <- intersect(target_genes, rownames(M))
  tot <- Matrix::colSums(M)
  tgt <- if (length(present)) Matrix::colSums(M[present, , drop = FALSE]) else rep(0, length(tot))
  setNames(ifelse(tot > 0, 100 * tgt / tot, 0), colnames(M))
})

# combine and align to all cells in object
pct_vec <- do.call(c, pct_list)
pct_vec <- pct_vec[colnames(d0_data)]
hist(pct_vec,
     breaks = 30,
     main = "Percent of UMIs in target genes",
     xlab = "Percent of UMIs",
     ylab = "Number of barcodes",
     cex.main = 3,
     cex.lab = 1.5
     )
abline(v = median(pct_vec), col = "red", lwd = 2)
# Label the median on the plot
text(median(pct_vec) + 12, 60000, labels = paste0("Median = ", round(median(pct_vec), 2)), col = "red", cex = 1.5)

d0_data$percent_target_umis <- pct_vec

hist(d0_data$nFeature_RNA, breaks = 30)
FeatureScatter(d0_data, feature1 = "nFeature_RNA", feature2 = "percent_target_umis", group.by = "orig.dose.chip")



p = FeatureScatter(d0_data, feature1 = "nCount_RNA", feature2 = "percent_target_umis")
p + geom_hline(yintercept = 60, color = "red", linetype = "dashed") + geom_vline(xintercept = 500, color = "blue", linetype = "dashed")
table(d0_data$percent_target_umis < 60 & d0_data$nCount_RNA > 500)


temp = subset(d0_data,subset = percent_target_umis >= 60)
counts <- rev(table(temp$orig.ident))

op <- par(mar = c(4, 15, 2, 1))  # more room on the left for long labels
xmax <- max(counts)

bp <- barplot(
  counts,
  horiz = TRUE,
  las = 1,              
  xlab = "Number of barcode",
  cex.names = 1.5,
  xlim = c(0, xmax * 1.12),
  main = "Barcodes passing filter on percent target UMIs >= 60%"
)
text(x = counts, y = bp, labels = counts, pos = 4, cex = 1.5, xpd = NA, offset = 0.5)

par(op)

d0_data_filtered = subset(d0_data, subset = percent_target_umis >= 60)
```

### Pseudobulk-based QC

```{r}
# Create pseudobulk based on the chip channels
bulk = Seurat::AggregateExpression(d0_data_filtered, group.by = "orig.dose.chip", assays = "RNA")

# Normalize to CPM
bulk_cpm = sweep(bulk$RNA, 2, colSums(bulk$RNA), "/") * 1e6

# Remove genes with 0 expression across all samples
bulk_cpm = bulk_cpm[rowSums(bulk_cpm) > 0, ]

genes_type_bulk <- setNames(genes_metadata$type, genes_metadata$Name)
present_bulk <- intersect(genes_metadata$Name, rownames(bulk_cpm))
# Get the list of genes for which there is no expression
missing_bulk <- setdiff(unique(genes_metadata$Name), present_bulk)
table(genes_type[missing_bulk])

bulk_cpm_matrix = as.matrix(bulk_cpm)

# Boxplot for the targeted genes vs non-targeted genes bulk CPM
dim(bulk_cpm_matrix)
targeted_genes = present_bulk
length(targeted_genes)
non_targeted_genes = setdiff(rownames(bulk_cpm_matrix), target_genes)
length(non_targeted_genes)

length(c(targeted_genes, non_targeted_genes))


tgt <- as.numeric(bulk_cpm_matrix[targeted_genes, , drop = FALSE])
nontgt <- as.numeric(bulk_cpm_matrix[non_targeted_genes, , drop = FALSE])

# If you want a log y-axis, zeros are not allowed â€” either drop zeros:
tgt_nz <- tgt[tgt > 0]
nontgt_nz <- nontgt[nontgt > 0]

boxplot(
  list("Targeted genes" = tgt_nz, "Non-targeted genes" = nontgt_nz),
  main = "Pseudobulk CPM distribution",
  ylab = "CPM",
  log = "y",
  outline = FALSE
)
stripchart(list("Targeted genes" = tgt_nz, "Non-targeted genes" = nontgt_nz), vertical = TRUE, method = "jitter", add = TRUE, pch = 20, col = rgb(0, 0, 0, 0.3))

# PCA analysis
pca = prcomp(t(bulk_cpm_matrix), scale. = TRUE)
plot(pca$x[,1], pca$x[,2], pch=16, cex=2, xlab=paste0("PC1: ",round(100*summary(pca)$importance[2,1],2),"%"), ylab=paste0("PC2:", round(100*summary(pca)$importance[2,2],2),"%"), main="PCA on pseudobulk CPM")
text(pca$x[,1], pca$x[,2], labels = rownames(pca$x), pos=3)

# Repeat PCA but only for targeted genes
pca_tgt = prcomp(t(bulk_cpm_matrix[targeted_genes, , drop = FALSE]), scale. = TRUE)
plot(pca_tgt$x[,1], pca_tgt$x[,2], pch=16, cex=2, xlab=paste0("PC1: ",round(100*summary(pca_tgt)$importance[2,1],2),"%"), ylab=paste0("PC2:", round(100*summary(pca_tgt)$importance[2,2],2),"%"), main="PCA on pseudobulk CPM (targeted genes only)")
text(pca_tgt$x[,1], pca_tgt$x[,2], labels = rownames(pca_tgt$x), pos=3)


# Check correlation between samples for the D0_markers from the genes_metadata type
d0_markers = genes_metadata$Name[genes_metadata$type == "D0_marker"]
length(d0_markers)
present_d0_markers <- intersect(d0_markers, rownames(bulk_cpm_matrix))
missing_d0_markers <- setdiff(unique(d0_markers), present_d0_markers)
length(missing_d0_markers)
table(genes_type[missing_d0_markers])
cor_matrix = cor(bulk_cpm_matrix[present_d0_markers, ], method = "pearson")
plot(cor_matrix)
```

Use ENSEMBLE ID as rownames

```{r}
x_names <- Read10X_h5(h5_files[1], use.names = TRUE)   # rownames = gene symbols
x_ids   <- Read10X_h5(h5_files[1], use.names = FALSE)  # rownames = Ensembl IDs

# Handle single-modality vs multi-modality (list)
pick_mod <- function(x, mod = c("Gene Expression","RNA")) {
  if (inherits(x, "dgCMatrix")) return(x)
  nms <- names(x); i <- match(tolower(mod), tolower(nms)); i <- i[!is.na(i)][1]
  if (is.na(i)) stop("Could not find RNA modality. Found: ", paste(nms, collapse=", "))
  x[[i]]
}

rna_names <- pick_mod(x_names)
rna_ids   <- pick_mod(x_ids)

# Build a mapping data.frame
gene_map <- data.frame(
  symbol  = rownames(rna_names),
  ensembl = rownames(rna_ids),
  stringsAsFactors = FALSE
)

rownames(d0_data_filtered)
d0_data_filtered[["RNA"]]@meta.data$ensembl = gene_map$ensembl[match(rownames(d0_data_filtered), gene_map$symbol)]
d0_data_filtered[["RNA"]]@meta.data$symbol = rownames(d0_data_filtered)
rownames(d0_data_filtered) = d0_data_filtered[["RNA"]]@meta.data$ensembl

```


### Sceptre analysis

```{r}
# Prepare the data for sceptre
#grna_target_df = read.table("metadata/gRNA-target.tsv", header = TRUE, sep = "\t")
grna_target_df = read.table(paste0(pairs_folder,"gRNA-target.tsv"), header = TRUE, sep = "\t")
head(grna_target_df)
colnames(grna_target_df) = c("grna_id", "grna_target")
grna_target_df$grna_id =  gsub("_", "-", grna_target_df$grna_id)
discovery_pairs = read.table(paste0(pairs_folder,"discovery_pairs_noposctrl.tsv"), header = TRUE, sep = "\t")
head(discovery_pairs)
positive_controls = read.table(paste0(pairs_folder,"positive_pairs.tsv"), header = FALSE, sep = "\t")
colnames(positive_controls) = c("grna_target", "response_id")
head(positive_controls)

d0_rna_assay = JoinLayers(d0_data_filtered[["RNA"]], new.layer.name = "counts", layers = "counts")
rna_counts <- GetAssayData(d0_rna_assay, assay = "RNA", layer = "counts")
grna_counts <- GetAssayData(d0_data_filtered, assay = "gRNA", layer = "counts")
grna_counts <- grna_counts[grna_target_df$grna_id, , drop = FALSE]
sceptre_object = sceptre::import_data(response_matrix = rna_counts[gene_map$ensembl[match(present, gene_map$symbol)], , drop = FALSE],
                                      grna_matrix = grna_counts,
                                      grna_target_data_frame = grna_target_df,
                                      moi = "high"
                                      )
sceptre_object
```

Quick clean-up

```{r}
rm(seurat_objects_list)
rm(d0_rna_assay)
rm(temp)
rm(d0_data)
rm(rna_counts)
rm(grna_counts)
rm(x_ids)
rm(x_names)
rm(rna_ids)
rm(rna_names)
gc()
```

QC and assign guides

```{r}
discovery_pairs = discovery_pairs[discovery_pairs$response_id %in% gene_map$ensembl[match(present, gene_map$symbol)], ]
positive_controls = positive_controls[positive_controls$response_id %in% gene_map$ensembl[match(present, gene_map$symbol)], ]

# Analysis parameters
sceptre_object <- set_analysis_parameters(
  sceptre_object = sceptre_object,
  discovery_pairs = discovery_pairs,
  positive_control_pairs = positive_controls,
  side = "both"
)

#sceptre_object <- assign_grnas(sceptre_object = sceptre_object, method = "mixture", parallel = TRUE, n_processors = 8)
sceptre_object <- assign_grnas(sceptre_object = sceptre_object,
                               method = "thresholding",
                               threshold = 10,
                               parallel = TRUE,
                               n_processors = 8)
plot(sceptre_object)
sceptre_object <- run_qc(sceptre_object, p_mito_threshold = 0.2)
print(sceptre_object)
plot(sceptre_object)
#sceptre_object <- assign_grnas(sceptre_object = sceptre_object, parallel = TRUE, n_processors = 8)

```


```{r}
sceptre_object <- run_power_check(sceptre_object, parallel = TRUE, n_processors = 2)
print(sceptre_object) # output suppressed for brevity
plot(sceptre_object)

```


Calibration check

```{r}
sceptre_object <- run_calibration_check(sceptre_object,
                                        parallel = TRUE,
                                        n_processors = 8,
                                        print_progress = TRUE,
                                        )
print(sceptre_object)
plot(sceptre_object)
```


```{r}
RM <- sceptre_object@response_matrix[[1]]   # expect a dgCMatrix
t = colnames(d0_data_filtered)
cells_in_use = t[sceptre_object@cells_in_use]

cell_chip <- d0_data_filtered@meta.data[cells_in_use, "orig.dose.chip", drop = TRUE]

idx_by_chip <- split(seq_along(cell_chip), cell_chip)

tags <- gsub("[^A-Za-z0-9._-]+", "_", names(idx_by_chip), perl = TRUE)

# 2) Write out one file per chip
outdir <- "response_matrices_by_dosechip"
dir.create(outdir, showWarnings = FALSE)

i <- 0L
for (nm in names(idx_by_chip)) {
  i <- i + 1L
  idx <- idx_by_chip[[nm]]
  M <- RM[, idx, drop = FALSE]         # much less overhead than doing which() each time
  saveRDS(M, file.path(outdir, paste0("RNA_response_", tags[i], ".rds")),
          compress = FALSE)            # fastest I/O; use compress=TRUE if size matters
}
```


## Results by 10X chip and dose

```{r}
# Run sceptre for each chip and dose combination
dose.chip.samples = unique(d0_data_filtered$orig.dose.chip)

# Save all the objects in a list
sceptre_objects_list = list()
for (sample in dose.chip.samples[1]){
  message("Processing sample: ", sample)
  temp = subset(d0_data_filtered, subset = orig.dose.chip == sample)
  d0_rna_assay = JoinLayers(temp[["RNA"]], new.layer.name = "counts")
  rna_counts <- GetAssayData(d0_rna_assay, assay = "RNA", layer = "counts")
  grna_counts <- GetAssayData(temp, assay = "gRNA", layer = "counts")
  grna_counts <- grna_counts[grna_target_df$grna_id, , drop = FALSE]
  # Update rownames with ENSEMBL IDs
  
  sceptre_object = sceptre::import_data(response_matrix = rna_counts[gene_map$ensembl[match(present, gene_map$symbol)], , drop = FALSE],
                                        grna_matrix = grna_counts,
                                        grna_target_data_frame = grna_target_df,
                                        moi = "high"
                                        )
  # Keep in the discovery and positive pairs only the genes present in the matrix
  discovery_pairs = discovery_pairs[discovery_pairs$response_id %in% gene_map$ensembl[match(present, gene_map$symbol)], ]
  positive_controls = positive_controls[positive_controls$V2 %in% gene_map$ensembl[match(present, gene_map$symbol)], ]
  sceptre_object <- set_analysis_parameters(
    sceptre_object = sceptre_object,
    discovery_pairs = discovery_pairs,
    positive_control_pairs = positive_controls,
    side = "both"
  )
  sceptre_object <- assign_grnas(sceptre_object = sceptre_object, parallel = TRUE, n_processors = 4)
  sceptre_objects_list[[sample]] = sceptre_object
}


```


