---
title: "00-create-sceptre-object"
output: html_document
editor_options: 
  chunk_output_type: console
---

# hiPSC -> EC

```{r}
library(hdf5r)
library(dplyr)
library(sceptre)
library(Seurat)
library(ggplot2)
library(ggtext)

# Custom functions
source("scripts/R/utils.R")
```

```{r}
# Create function to save ggplot
save_ggplot <- function(plot, filename, width = 8, height = 6, dpi = 300) {
  ggsave(
    filename = filename,
    plot = plot,
    width = width,
    height = height,
    dpi = dpi
  )
}
```

## Parameters and file paths
```{r}
# INPUTS
metadata_samples_fnp <- "metadata/metadata_sample_endothelial_terra.tsv"
metadata_genes_fnp <- "metadata/metadata_genes_final.csv"
metadata_guides_fnp <- "metadata/guide_metadata_with_new_accessions.tsv"
discovery_pairs_fnp <- "results/03-2025-10-24_create_sceptre_input/discovery_pairs_no_pos_hg38_readout_genes.tsv"
positive_pairs_fnp <- "results/03-2025-10-24_create_sceptre_input/positive_pairs_hg38_readout_genes.tsv"
negative_controls_fnp <- "results/03-2025-10-24_create_sceptre_input/sceptre_non_targeting_guides_hg38.tsv"

# OUTPUTS
output_folder <- "results/04-2025-10-27_sceptre_analysis"
```

## Load the metadata
```{r}
metadata_samples_df <- read.csv(metadata_samples_fnp, sep = "\t")
metadata_samples_df

metadata_genes_df <- read.csv(metadata_genes_fnp)
metadata_genes_df

target_genes <- metadata_genes_df$intended_target_gene_symbol
target_genes <- unique(target_genes) # Because we have multiple primers for the same gene

h5_files_fnp <- list.files(path = "data/h5", pattern = "*.h5", full.names = TRUE, recursive = TRUE)

h5_sample_names <- sub(".*/(.*)_sample_filtered_feature_bc_matrix\\.h5$", "\\1", h5_files_fnp)

h5_sample_names
```

## Create the gene symbol to ensembl ID mapping
```{r}
# Load the same h5 twice, one time using gene symbols and one time using Ensembl IDs
temp_seurat <- Seurat::Read10X_h5(h5_files_fnp[1], use.names = TRUE)   # rownames = gene symbol
gene_symbol_from_h5 <-  rownames(temp_seurat[[1]])
head(gene_symbol_from_h5)

temp_seurat <- Seurat::Read10X_h5(h5_files_fnp[1], use.names = FALSE)  # rownames = Ensembl IDs
ensembl_id_from_h5 <- rownames(temp_seurat[[1]])
head(ensembl_id_from_h5)

# Create a mapping data.frame
gene_map <- data.frame(
  symbol = gene_symbol_from_h5,
  ensembl = ensembl_id_from_h5,
  stringsAsFactors = FALSE
)
head(gene_map)
rm(temp_seurat, gene_symbol_from_h5, ensembl_id_from_h5)
```

## Read the count matrices in Seurat
```{r}
#Keeping only day 0 for speed
#h5_sample_names = metadata_samples_df$entity.endothelial_id[ metadata_samples_df$day == 0]

#Keeping D0 and D2 and make it general
h5_sample_names = metadata_samples_df$entity.endothelial_id[ metadata_samples_df$day == 0 | metadata_samples_df$day == 2]
h5_files_fnp = h5_files_fnp[basename(h5_files_fnp) %in% paste0(h5_sample_names,"_sample_filtered_feature_bc_matrix.h5")]

seurat_objects_list <- mapply(
  read_10x_rna_grna,
  h5 = h5_files_fnp,
  sample_name = h5_sample_names,
  SIMPLIFY = FALSE
)

names(seurat_objects_list) <- h5_sample_names

d0_data <- merge(x = seurat_objects_list[[1]], 
                 y = seurat_objects_list[2:length(seurat_objects_list)],
                 merge.data = TRUE,
                 add.cell.ids = h5_sample_names)
Assays(d0_data)
d0_data@meta.data$all = "Combined"
```

# Add metadata
```{r}

ids <- Cells(d0_data)
sample_id <- sub("_[^_]+$", "", ids)

meta <- metadata_samples_df
meta$sample_id <- meta$entity.endothelial_id
meta <- meta[!duplicated(meta$sample_id), , drop = FALSE]

idx <- match(sample_id, meta$sample_id)

meta_for_seurat <- meta[idx, , drop = FALSE]
meta_for_seurat$sample_id <- NULL
rownames(meta_for_seurat) <- ids


d0_data <- AddMetaData(d0_data, metadata = meta_for_seurat)
```


```{r}
# clean up
rm(seurat_objects_list)
gc()
```

## QC the RNAseq

### Check how many target genes are present in the dataset
```{r}
n_readout_genes_in_matrix <- length(intersect(rownames(d0_data), target_genes))
message("All genes present in the dataset: ", n_readout_genes_in_matrix, "\n")
```

### Update metadata
```{r}
d0_data$replicate_factor = factor(
    meta_for_seurat$replicate, 
    
    # The actual values in the data (the numeric levels)
    levels = 1:4, 
    
    # The new labels to assign to those levels (the string names)
    labels = c("Rep 1", "Rep 2", "Rep 3", "Rep 4")
)

d0_data$day_factor = factor(
    meta_for_seurat$day, 
    
    # The actual values in the data (the numeric levels)
    levels = c(0,2,4), 
    
    # The new labels to assign to those levels (the string names)
    labels = c("Day 0", "Day 2", "Day 4")
)
```


### Compute percentage or mitochondrial UMIs
```{r}
# Compute percentage of Mitochondrial genes
d0_data[["percent.mt"]] <- Seurat::PercentageFeatureSet(d0_data, pattern = "^MT-", assay = "RNA")
# Plot mitochondrial percentage
median_stats <- as.data.frame(d0_data@meta.data %>%
  group_by(day_factor) %>%
  summarise(
    median_val = median(percent.mt, na.rm = TRUE),
    max_count = max(hist(percent.mt, plot=FALSE)$counts) 
  ))

p <- ggplot(d0_data@meta.data, aes(x = percent.mt)) +
  geom_histogram(bins = 50, color = "black", position = "identity", alpha = 0.6) +
  geom_vline(data = median_stats, 
             aes(xintercept = median_val), 
             linetype = "dashed",
             color = "red", 
             linewidth = 1) +
  geom_text(data = as.data.frame(median_stats),
            aes(x = median_val, 
                # Place text slightly above the top of the histogram 
                y = max(ggplot_build(ggplot(d0_data@meta.data, aes(x=percent.mt)) + geom_histogram(bins=50, position="identity"))$data[[1]]$count) * 0.5,
                label = paste0("Median: ", round(median_val, 2))),
            # Nudge the label slightly to the right to avoid collision with the line
            hjust = -0.5,
            vjust = 0.5, 
            show.legend = FALSE,
            color = "black") + 
  labs(title = "Mitochondrial Percentage Distribution",
       x = "UMIs from Mitochondrial genes(%)",
       y = "Barcode Count") +
  facet_grid(rows = vars(day_factor)) +
  theme_minimal(base_size = 14)

save_ggplot(plot = p, filename = paste0(output_folder,"/plots/mitochondrial_percent.png"))
save_ggplot(plot = p, filename = paste0(output_folder,"/plots/mitochondrial_percent.pdf"))
```

### Compute percentage of ribosomal UMIs
```{r}
# Compute percentage of Ribosomal genes
d0_data[["percent.ribo"]] <- Seurat::PercentageFeatureSet(d0_data, pattern = "^RPS|^RPL|^RRP|^MRP|^FAU", assay = "RNA")
# Plot ribosomal percentage
median_stats <- as.data.frame(d0_data@meta.data %>%
  group_by(day_factor) %>%
  summarise(
    median_val = median(percent.ribo, na.rm = TRUE),
    max_count = max(hist(percent.ribo, plot=FALSE)$counts) 
  ))

p <- ggplot(d0_data@meta.data, aes(x = percent.ribo)) +
  geom_histogram(bins = 50, color = "black", position = "identity", alpha = 0.6) +
  geom_vline(data = median_stats, 
             aes(xintercept = median_val), 
             linetype = "dashed",
             color = "red", 
             linewidth = 1) +
  geom_text(data = as.data.frame(median_stats),
            aes(x = median_val, 
                # Place text slightly above the top of the histogram 
                y = max(ggplot_build(ggplot(d0_data@meta.data, aes(x=percent.ribo)) + geom_histogram(bins=50, position="identity"))$data[[1]]$count) * 0.5,
                label = paste0("Median: ", round(median_val, 2))),
            # Nudge the label slightly to the right to avoid collision with the line
            hjust = -0.5,
            vjust = 0.5, 
            show.legend = FALSE,
            color = "black") + 
  labs(title = "Ribosomal Percentage Distribution",
       x = "UMIs from Ribosomal genes(%)",
       y = "Barcode Count") +
  facet_grid(rows = vars(day_factor)) +
  theme_minimal(base_size = 14)

save_ggplot(plot = p, filename = paste0(output_folder,"/plots/ribosomal_percent.png"))
save_ggplot(plot = p, filename = paste0(output_folder,"/plots/ribosomal_percent.pdf"))
```

### Compute percent of UMIs in readout genes
```{r}
# Get the layer names
layers <- grep("^counts", Layers(d0_data[["RNA"]]), value = TRUE)
# Iterate and compute percent of UMIs in target genes
pct_list <- lapply(layers, function(ly) {
  M <- SeuratObject::GetAssayData(d0_data, assay = "RNA", layer = ly)
  present <- intersect(target_genes, rownames(M))
  tot <- Matrix::colSums(M)
  tgt <- if (length(present)) Matrix::colSums(M[present, , drop = FALSE]) else rep(0, length(tot))
  setNames(ifelse(tot > 0, 100 * tgt / tot, 0), colnames(M))
})

# Combine results
pct_vec <- do.call(c, pct_list)
pct_vec <- pct_vec[colnames(d0_data)]

# Add to metadata
d0_data$percent_target_umis <- pct_vec

# Plot ribosomal percentage
median_stats <- as.data.frame(d0_data@meta.data %>%
  group_by(day_factor) %>%
  summarise(
    median_val = median(percent_target_umis, na.rm = TRUE),
    max_count = max(hist(percent_target_umis, plot=FALSE)$counts) 
  ))

p <- ggplot(d0_data@meta.data, aes(x = percent_target_umis)) +
  geom_histogram(bins = 50, color = "black", position = "identity", alpha = 0.6) +
  geom_vline(data = median_stats, 
             aes(xintercept = median_val), 
             linetype = "dashed",
             color = "red", 
             linewidth = 1) +
  geom_text(data = as.data.frame(median_stats),
            aes(x = median_val, 
                # Place text slightly above the top of the histogram 
                y = max(ggplot_build(ggplot(d0_data@meta.data, aes(x=percent_target_umis)) + geom_histogram(bins=50, position="identity"))$data[[1]]$count) * 0.5,
                label = paste0("Median: ", round(median_val, 2))),
            # Nudge the label slightly to the right to avoid collision with the line
            hjust = -0.5,
            vjust = 0.5, 
            show.legend = FALSE,
            color = "black") + 
  labs(title = "Target Genes Percentage Distribution",
       x = "UMIs from targeted genes(%)",
       y = "Barcode Count") +
  facet_grid(rows = vars(day_factor)) +
  theme_minimal(base_size = 14)

save_ggplot(plot = p, filename = paste0(output_folder,"/plots/target_genes_umis_percent.png"))
save_ggplot(plot = p, filename = paste0(output_folder,"/plots/target_genes_umis_percent.pdf"))


# Plot number of UMIs vs percent target UMIs
# I want to color the on the x-axis the threshold
x_breaks <- seq(0, 30000, by = 5000)
x_breaks <- sort(unique(c(x_breaks, 1000)))
x_labels <- as.character(x_breaks)
x_labels[x_breaks == 1000] <- "<span style='color:blue'>1000</span>"

# Samples combined
p<-Seurat::FeatureScatter(
  d0_data,
  feature1 = "nCount_RNA",
  feature2 = "percent_target_umis",
  group.by = "all",
  split.by = "day_factor"
) +
  ggtitle("Total UMIs vs UMIs in readout genes", subtitle = paste("n = ", ncol(d0_data))) +
  geom_hline(yintercept = 50, linetype = "dashed", color = "red") +
  geom_vline(xintercept = 1000, linetype = "dashed", color = "blue") +
  scale_x_continuous(
    breaks = x_breaks,
    labels = x_labels,
    limits = c(0, 30000)
  ) +
  scale_y_continuous(
    breaks = c(0, 20, 40, 50, 60, 80, 100),
    labels = c("0", "20%", "40%", "<span style='color:red'>50%</span>", "60%", "80%", "100%"),
    limits = c(0, 100)
  ) +
  scale_color_manual(name="Replicate", values = "black") +
  xlab("Total UMIs") +
  ylab("UMIs in readout genes") +
  theme(
    plot.title = element_text(hjust = 0.5),
    axis.text.x = ggtext::element_markdown(),
    axis.text.y = ggtext::element_markdown()
  )

save_ggplot(plot = p, filename = paste0(output_folder,"/plots/total_umi_vs_target_genes_umis_percent.png"))
save_ggplot(plot = p, filename = paste0(output_folder,"/plots/total_umi_vs_target_genes_umis_percent.pdf"))

# Are the barcodes with low ribosomal percentage the one we want to remove
p <- Seurat::FeatureScatter(
  d0_data[,d0_data$percent.ribo < 20],
  feature1 = "nCount_RNA",
  feature2 = "percent_target_umis",
  group.by = "all",
  split.by = "day_factor"
) +
  ggtitle("Total UMIs vs UMIs in readout genes", subtitle = paste0("Ribosomal percentage < 20. n = ",ncol(d0_data[,d0_data$percent.ribo < 20]))) +
  geom_hline(yintercept = 50, linetype = "dashed", color = "red") +
  geom_vline(xintercept = 1000, linetype = "dashed", color = "blue") +
  scale_x_continuous(
    breaks = x_breaks,
    labels = x_labels,
    limits = c(0, 30000)
  )+
  scale_y_continuous(
    breaks = c(0, 20, 40, 50, 60, 80, 100),
    labels = c("0", "20%", "40%", "<span style='color:red'>50%</span>", "60%", "80%", "100%"),
    limits = c(0, 100)
  ) +
  scale_color_manual(name="Replicate", values = "black") +
  xlab("Total UMIs") +
  ylab("UMIs in readout genes") +
  theme(
    plot.title = element_text(hjust = 0.5),
    axis.text.x = ggtext::element_markdown(),
    axis.text.y = ggtext::element_markdown()
  )

save_ggplot(plot = p, filename = paste0(output_folder,"/plots/total_umi_vs_target_genes_umis_percent_ribo_filtered.png"))
save_ggplot(plot = p, filename = paste0(output_folder,"/plots/total_umi_vs_target_genes_umis_percent_ribo_filtered.pdf"))
```

```{r}
# Get general statistics of barcodes per channel, replicate
table(d0_data$chip, d0_data$batch)
table(d0_data$chip, d0_data$channel)
table(d0_data$replicate, d0_data$batch)
table(d0_data$replicate_factor, d0_data$day_factor)

```


### Compare CPM of readout genes vs non-readout genes
```{r}
# Compute CPM
layers <- grep("^counts", Layers(d0_data[["RNA"]]), value = TRUE)
# Iterate per layer
cpm_list <- lapply(layers, function(ly) {
  M <- SeuratObject::GetAssayData(d0_data, assay = "RNA", layer = ly)
  # Compute CPM
  tot <- Matrix::colSums(M)
  cpm <- t(t(M) / tot * 1e6)
  cpm
})
# Combine results
cpm_mat <- do.call(cbind, cpm_list)
# Subset to readout genes and non-readout genes
present <- intersect(target_genes, rownames(cpm_mat))
non_present <- setdiff(rownames(cpm_mat), target_genes)
cpm_readout <- cpm_mat[present, , drop = FALSE]
cpm_non_readout <- cpm_mat[non_present, , drop = FALSE]
# Compute mean CPM per gene
mean_cpm_readout <- Matrix::rowMeans(cpm_readout)
mean_cpm_non_readout <- Matrix::rowMeans(cpm_non_readout)
# Combine into a data.frame for plotting
cpm_df <- data.frame(
  gene = c(names(mean_cpm_readout), names(mean_cpm_non_readout)),
  mean_cpm = c(mean_cpm_readout, mean_cpm_non_readout),
  type = c(rep("readout", length(mean_cpm_readout)), rep("non-readout", length(mean_cpm_non_readout)))
)
# Plot
ggplot(cpm_df, aes(x = type, y = mean_cpm + 1)) +
  geom_boxplot(outlier.size = 0.5) +
  scale_y_log10() +
  ylab("Mean CPM (log10 scale)") +
  xlab("") +
  ggtitle("Mean CPM of readout genes vs non-readout genes") +
  theme(
    plot.title = element_text(hjust = 0.5),
  ) +
  theme_minimal()
```

### Check CPM of non-readout genes aggregated by the presence of inner primer
```{r}
# Number of non-readout genes with CPM >= 100 
high_cpm_non_readout_genes <- unique(cpm_df$gene[cpm_df$type == "non-readout" & cpm_df$mean_cpm >= 100])
# Add the ENSEMBL id for the genes and save
cpm_df$ensembl_id <- gene_map$ensembl[match(cpm_df$gene, gene_map$symbol)]
write.table(
  cpm_df[high_cpm_non_readout_genes,c("gene", "ensembl_id", "mean_cpm")],
  file = paste0(output_folder,"/high_cpm_non_readout_genes.tsv", sep=""),
  quote = FALSE,
  row.names = FALSE,
  col.names = c("gene_symbol", "ensembl_id", "mean_cpm"),
  sep = "\t"
  )
# These need to be fixed
# message("Number of non-readout genes with mean CPM >= 100: ",length(high_cpm_non_readout_genes), "\n")
# # Add inner primer information to cpm_df
# metadata_primers_df$is_intended_target <- as.logical(metadata_primers_df$is_intended_target)
# 
# non_readout_genes_with_primer <- unique(metadata_primers_df$gene_symbol[!metadata_primers_df$is_intended_target])
# message("Number of non-readout genes with inner primer: ", length(non_readout_genes_with_primer), "\n")
# 
# cpm_df$has_inner_primer <- ifelse(cpm_df$gene %in% non_readout_genes_with_primer, "with_inner_primer", "without_inner_primer")
# # Subset to non-readout genes
# cpm_non_readout_df <- cpm_df[cpm_df$type == "non-readout" & cpm_df$mean_cpm >= 100, ]
# # Plot
# ggplot(cpm_non_readout_df, aes(x = has_inner_primer, y = mean_cpm + 1)) +
#   geom_boxplot(outlier.size = 0.5) +
#   scale_y_log10() +
#   ylab("Mean CPM (log10 scale)") +
#   xlab("") +
#   ggtitle("Mean CPM of non-readout genes by inner primer presence") +
#   theme(
#     plot.title = element_text(hjust = 0.5),
#   ) +
#   theme_minimal()
```

### Number of genes per barcode
```{r}
# Plot genes per barcode with median annotated
hist(d0_data$nFeature_RNA, breaks = 50, main = "Number of genes per barcode", xlab = "Number of detected genes", ylab = "Number of barcodes")
abline(v = median(d0_data$nFeature_RNA, na.rm = TRUE), col = "red", lwd = 2, lty = 3)
legend_text <- paste0("Median: ", round(median(d0_data$nFeature_RNA, na.rm = TRUE), 2))
legend("topright", legend = legend_text, col = "red", lwd = 2, lty = 3, bty = "n")
```

### Number of UMIs vs number of genes
```{r}
# Plot number of UMIs vs number of genes
Seurat::FeatureScatter(
  d0_data,
  feature1 = "nCount_RNA",
  feature2 = "nFeature_RNA",
  group.by = "replicate_factor",
  split.by = "replicate_factor"
) +
  scale_color_manual(values = palette_4_qual[c(2,1,3,4)]) +
  xlab("Total UMIs") +
  ylab("Number of genes") +
  ggtitle("Total UMIs vs number of genes") +
  theme(
    plot.title = element_text(hjust = 0.5),
  )
```

## Filter cells
Using percentage of UMI from target genes >= 55%, percentage of ribosomal UMIs>=21, and number of UMIs >= 1500

```{r}
d0_data_filtered = subset(d0_data, subset = percent_target_umis >= 55 & nCount_RNA >= 1500 & percent.ribo >= 21)
d0_data_filtered
message("Number of cells before filtering: ", ncol(d0_data), "\n")
message("Number of cells filtered out: ", ncol(d0_data) - ncol(d0_data_filtered), "\n")
message("Number of cells after filtering: ", ncol(d0_data_filtered), "\n")
```

### Normalize the data
```{r}
d0_data_filtered_norm <- Seurat::NormalizeData(d0_data_filtered[["RNA"]], normalization.method = "RC", scale.factor = 1e6)
```

### Average number of CPM per gene from normalized data
```{r}
# This works with high memory machines
#d0_rna_assay_norm <- SeuratObject::JoinLayers(d0_data_filtered_norm, new.layer.name = "norm", layers = "data")
#mean_cpm_per_barcode <- Matrix::rowMeans(GetAssayData(d0_data_filtered_norm, assay = "RNA", layer = "data"))

## 1) Build a single CPM matrix by column-binding all data.* layers
data_layers <- grep("^data(\\.|$)", Layers(d0_data_filtered_norm), value = TRUE)
stopifnot(length(data_layers) > 0)

mats <- lapply(data_layers, function(ly) SeuratObject::GetAssayData(d0_data_filtered_norm, assay="RNA", layer = ly))
cpm <- do.call(cbind, mats)   # genes x cells, CPM units; each column sums to ~1e6


## A) Gene-wise mean CPM across ALL cells (including zeros):
mean_cpm_per_gene_allcells <- Matrix::rowMeans(cpm)

```


Save the mean CPM per gene
```{r}
# Add gene symbol and create output dataframe
mean_cpm_df <- data.frame(
  gene_symbol = names(mean_cpm_per_gene_allcells),
  ensembl_id = gene_map$ensembl[match(names(mean_cpm_per_gene_allcells), gene_map$symbol)],
  mean_cpm = mean_cpm_per_gene_allcells,
  is_target_gene = ifelse(names(mean_cpm_per_gene_allcells) %in% target_genes, TRUE, FALSE),
  stringsAsFactors = FALSE
)

write.table(
  mean_cpm_df,
  file = paste0(output_folder,"/mean_cpm_per_gene_filtered_cells.tsv", sep=""),
  quote = FALSE,
  row.names = FALSE,
  col.names = c("gene_symbol", "ensembl_id", "mean_cpm", "is_readout_gene"),
  sep = "\t"
)
```

### Save the Seurat object
```{r}
SeuratObject::SaveSeuratRds(
  d0_data_filtered_norm,
  file = paste0(output_folder,"/seurat_object_filtered_norm_day00.rds")
  )
```


```{r}
# # Faceted
# Seurat::FeatureScatter(
#   d0_data,
#   feature1 = "nCount_RNA",
#   feature2 = "percent_target_umis",
#   group.by = "replicate_factor",
#   split.by = "replicate_factor"
# ) +
#   ggtitle("Total UMIs vs UMIs in readout genes") +
#   geom_hline(yintercept = 50, linetype = "dashed", color = "red") +
#   geom_vline(xintercept = 1000, linetype = "dashed", color = "blue") +
#   scale_x_continuous(
#     breaks = x_breaks,
#     labels = x_labels
#   ) +
#   scale_y_continuous(
#     breaks = c(0, 20, 40, 50, 60, 80, 100),
#     labels = c("0", "20%", "40%", "<span style='color:red'>50%</span>", "60%", "80%", "100%")
#   ) +
#   scale_color_manual(name="Replicate", values = palette_4_qual[c(2,1,3,4)]) +
#   xlab("Total UMIs") +
#   ylab("UMIs in readout genes") +
#   theme(
#     plot.title = element_text(hjust = 0.5),
#     axis.text.x = ggtext::element_markdown(),
#     axis.text.y = ggtext::element_markdown(),
#   )
```

