# Align guides to the genome

## Generate FASTA from metadata table

I have to do it in two steps because the positive controls don't have the sequence with the PAM.
First all the guides without the positive controls:

```bash
 grep -v positive_control ../../metadata/hiPSC-HPC/All_guides_with_addnl_controls_altTSS.csv | cut -f 10,12 -d ',' | awk -v FS="," 'NR>1{print ">"$1"\n"$2}' > guides_no_pos_controls.fa
 ```

Then the positive controls. I need to add the leading G and the PAM (NGG):
```bash
grep positive_control ../../metadata/hiPSC-HPC/All_guides_with_addnl_controls_altTSS.csv | cut -f9,10 -d ',' | awk -v FS=',' '{print ">"$2"\nG"$1"NGG"}' > guide_pos_controls.fa
 ```

Then concatenate them:

```bash
 cat guides_no_pos_controls.fa pos_controls.fa > all_guides.fa
 ```

## Generate FASTQ from FASTA

```bash
awk -f ../../../broad-nnfc-dc-tap/src/awk/fasta2fastq.awk all_guides.fasta > all_guides.fastq

bgzip all_guides.fastq
```
### Align to the genome using GEM mapper

```bash
../../gem3-mapper/bin/gem-mapper \
  -I ../../annotations/ENCODE/hg38/gem_index/GRCh38_no_alt_analysis_set_GCA_000001405.15.gem \
  -i hpc_all_guides.fq.gz \
  -o hpc_all_guides_hg38.sam \
  --mapping-mode sensitive \
  --alignment-max-error 0.1 \
  --alignment-global-min-identity 0.9 \
  --threads 32
```

### Convert the SAM to BED

This is the SAM alignment file generated by the GEM mapper, which includes XA tags for alternative alignments.

```bash
python3 ../../../broad-nnfc-dc-tap/src/python/convert_sam_with_xa_to_bed.py all_guides_hg38.sam
```

### Merge the BED and obtain elements

```bash
grep primary all_guides_hg38.bed | sort -k1,1 -k2,2n | bedtools merge -i - -d 50 -c 4 -o distinct,count_distinct > targeted_elements_hg38.bed
```
Verify that one guide is present in only one element:

```bash
grep primary all_guides_hg38.bed | bedtools intersect -a - -b targeted_elements_hg38.bed -wa -wb | cut -f4,8 | sort | uniq -c | awk '$1>1'
```
This command should return no output. If it does, it means that at least one guide is present in more than one element.


### Create gRNA input for SCEPTRE

```bash
awk -v OFS="\t" '{split($4,v,","); for (n=1; n<=length(v); n++){print v[n],$1":"$2"-"$3}}' targeted_elements_hg38.bed > grna_sceptre_input.tsv
```

Adding non-targeting guides:

```bash
awk -v OFS="\t" 'NR>1{print $1,"non-targeting"}' all_guides_hg38_unmapped.tsv >> grna_sceptre_input.tsv
```

### Annotate TSS

```bash
gzip -dc ~/Annotations/HG38/GENCODE/gencode.v44.basic.annotation.gtf.gz | awk 'OFS="\t" {if ($3=="transcript") {if ($7 == "+") {print $1,$4-1,$4,$12,".",$7, $10, $14,$16} else {print $1,$5-1,$5,$12,".",$7,$10,$14,$16}}}' | tr -d '";' | sort -k1,1V -k2,2n > gencode_tss.bed
```


### Re-annotating readout genes

Find mismatching readout genes:
```bash
awk 'NR>1' primer_matches.csv | cut -d ',' -f 2,6,7  | sort -u |
awk -F, '{
  k=$1; gsub(/^ +| +$/,"",k);           # trim spaces just in case
  v=tolower($3); gsub(/^ +| +$/,"",v);
  n[k]++; if (v=="true") t[k]++
}
END {
  for (k in n) if (t[k]==0) print k
}' > missing_genes.txt 
```

```bash
grep -wFf missing_genes.txt primer_matches.csv | cut -d ',' -f 2,8,10,11 | sort -u > map_missing_v32_genes_to_v44.csv
```
